File: mobile/library/swift/src/mocks/MockStreamPrototype.swift
import Dispatch
@_implementationOnly import EnvoyEngine
import Foundation

/// Mock implementation of `StreamPrototype` which is used to produce `MockStream` instances.
@objcMembers
public final class MockStreamPrototype: StreamPrototype {
  private let onStart: ((MockStream) -> Void)?

  /// Initialize a new instance of the mock prototype.
  ///
  /// - parameter onStart: Closure that will be called each time a new stream
  ///                      is started from the prototype.
  init(onStart: @escaping (MockStream) -> Void) {
    self.onStart = onStart
    super.init(engine: MockEnvoyEngine())
  }

  public override func start(queue: DispatchQueue = .main) -> Stream {
    let callbacks = self.createCallbacks(queue: queue)
    let stream = MockStream(mockStream: MockEnvoyHTTPStream(handle: 0, callbacks: callbacks))
    self.onStart?(stream)
    return stream
  }
}

File: mobile/test/kotlin/io/envoyproxy/envoymobile/mocks/MockStreamPrototype.kt
package io.envoyproxy.envoymobile.mocks

import io.envoyproxy.envoymobile.Stream
import io.envoyproxy.envoymobile.StreamPrototype
import java.util.concurrent.Executor

/**
 * Mock implementation of `StreamPrototype` which is used to produce `MockStream` instances.
 *
 * @param onStart Closure that will be called each time a new stream is started from the prototype.
 */
class MockStreamPrototype(private val onStart: ((stream: MockStream) -> Unit)?) :
  StreamPrototype(MockEnvoyEngine()) {
  override fun start(executor: Executor?): Stream {
    val callbacks = createCallbacks(executor)
    val stream = MockStream(MockEnvoyHTTPStream(callbacks, false))
    onStart?.invoke(stream)
    return stream
  }
}
